app/models/host_aspect_base.rb

```diff
@@ -0,0 +1,36 @@
+class HostAspectBase < ActiveRecord::Base
+  self.abstract_class = true
+
+  def self.inherited(subclass)
+    super
+    subclass.class_eval do
+      belongs_to_host
+    end
+  end
+
+  def info
+    {}
+  end
+
+  def smart_proxy_ids
+    []
+  end
+
+  def after_clone
+  end
+
+  def template_filter_options(kind)
+    {}
+  end
+
+  def provisioning_template_options(base_options)
+    base_options
+  end
+
+  def self.inherited_attributes(hostgroup, aspect_attributes)
+    aspect_attributes
+  end
+
+  def self.populate_fields_from_facts(host, importer, type, proxy_id)
```

Comment:
That's right, but it won't work the other way around - if you try to call the method while specifying proxy on something other than Host::Managed.
Maybe the better option will be to create some kind of context object that will be populated by the host, and then passed to this method and the whole chain that it is calling. This context object can be a simple options hash.